daemons/attrd/attrd_cib.c

```diff
@@ -301,48 +302,98 @@ attrd_cib_callback(xmlNode *msg, int call_id, int rc, xmlNode *output, void *use
+ * \param[in] set_type  Type of attribute set block (\c XML_TAG_ATTR_SETS or
+ *                      \c XML_TAG_UTILIZATION)
+ * \param[in] nodeid    ID of node for which to update attribute value
+ * \param[in] value     New value for attribute
+ *
+ * \return Legacy Pacemaker return code
+ */
+static int
+add_attr_update(const attribute_t *attr, const char *set_type,
+                const char *nodeid, const char *value)
+{
+    char *set_id = NULL;
+    char *attr_id = NULL;
+    int rc = pcmk_ok;
 
-    if (pcmk__str_eq(a->set_type, XML_TAG_ATTR_SETS, pcmk__str_null_matches)) {
-        xml_obj = create_xml_node(xml_obj, XML_TAG_ATTR_SETS);
-    } else if (pcmk__str_eq(a->set_type, XML_TAG_UTILIZATION, pcmk__str_none)) {
-        xml_obj = create_xml_node(xml_obj, XML_TAG_UTILIZATION);
+    if (attr->set_id != NULL) {
+        pcmk__str_update(&set_id, attr->set_id);
     } else {
-        crm_err("Unknown set type attribute: %s", a->set_type);
+        set_id = crm_strdup_printf("%s-%s", XML_CIB_TAG_STATUS, nodeid);
     }
+    crm_xml_sanitize_id(set_id);
 
-    if (a->set_id) {
-        crm_xml_set_id(xml_obj, "%s", a->set_id);
+    if (attr->uuid != NULL) {
+        pcmk__str_update(&attr_id, attr->uuid);
     } else {
-        crm_xml_set_id(xml_obj, "%s-%s", XML_CIB_TAG_STATUS, nodeid);
+        attr_id = crm_strdup_printf("%s-%s", set_id, attr->id);
     }
-    set = ID(xml_obj);
+    crm_xml_sanitize_id(attr_id);
 
-    xml_obj = create_xml_node(xml_obj, XML_CIB_TAG_NVPAIR);
-    if (a->uuid) {
-        crm_xml_set_id(xml_obj, "%s", a->uuid);
-    } else {
-        crm_xml_set_id(xml_obj, "%s-%s", set, a->id);
-    }
-    crm_xml_add(xml_obj, XML_NVPAIR_ATTR_NAME, a->id);
+    if (value != NULL) {
+        xmlNode *update = create_xml_node(NULL, XML_CIB_TAG_STATE);
+        xmlNode *child = update;
+
+        rc = -ENOMEM;   // For any failed create_xml_node() call
+        if (child == NULL) {
+            goto done;
+        }
+        crm_xml_add(child, XML_ATTR_ID, nodeid);
+
+        child = create_xml_node(child, XML_TAG_TRANSIENT_NODEATTRS);
+        if (child == NULL) {
+            goto done;
+        }
+        crm_xml_add(child, XML_ATTR_ID, nodeid);
+
+        child = create_xml_node(child, set_type);
+        if (child == NULL) {
+            goto done;
+        }
+        crm_xml_add(child, XML_ATTR_ID, set_id);
+
+        child = create_xml_node(child, XML_CIB_TAG_NVPAIR);
+        if (child == NULL) {
+            goto done;
+        }
+        crm_xml_add(child, XML_ATTR_ID, attr_id);
+        crm_xml_add(child, XML_NVPAIR_ATTR_NAME, attr->id);
+        crm_xml_add(child, XML_NVPAIR_ATTR_VALUE, value);
 
-    if(value) {
-        crm_xml_add(xml_obj, XML_NVPAIR_ATTR_VALUE, value);
+        rc = cib_internal_op(the_cib, PCMK__CIB_REQUEST_MODIFY, NULL,
+                             XML_CIB_TAG_STATUS, update, NULL,
+                             cib_can_create|cib_transaction, attr->user);
+        free_xml(update);
```

Comment:
This needs to happen on failure too. `update` needs a broader scope so that we can free it in `done`.